pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+			T::Bank::transfer(&id, &T::Bank::treasury()?, T::AuctionStartFee::get())?;
+
+			Self::deposit_event(Event::<T>::AuctionCreated { who: id, auction_id });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::bid())]
+		pub fn bid(
+			origin: OriginFor<T>,
+			auction_id: AuctionId,
+			price: T::Balance,
+		) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// update storage auctions
+			Auctions::<T>::try_mutate(auction_id, |maybe_auction_data| {
+				if let Some(auction_data) = maybe_auction_data.as_mut() {
+					let current_price = sp_std::cmp::max(
+						if let Some((last_bidder, last_price)) = &auction_data.current_bid {
+							// Refund the previous bidder's bid
+							T::Bank::transfer(
+								&T::BidsPoolAccount::get(),
+								last_bidder,
+								*last_price,
+							)?;
+							Ok::<T::Balance, DispatchError>(*last_price)
+						} else {
+							Ok(Default::default())
+						}?,
+						auction_data.start.unwrap_or_default(),
+					);
+
+					// Ensure bid is valid.
+					ensure!(
+						price >= current_price + T::MinimumIncrease::get(),
+						Error::<T>::BidPriceTooLow
+					);
+
+					// Transfer bid to Bids Pool's account.
+					T::Bank::transfer(&id, &T::BidsPoolAccount::get(), price)?;
+
+					// Update current bid's storage.
+					auction_data.current_bid = Some((id.clone(), price));
+
+					// Calculate how many blocks the auction will over, if it shorter than specific
+					// length, then extend to specific length.
+					if T::ExtendedLength::get() >
+						auction_data.expiry_block -
+							frame_system::Pallet::<T>::current_block_number()
+					{
+						// Remove the old auction id
+						AuctionsExpiryBlock::<T>::mutate(auction_data.expiry_block, |auctions| {
+							auctions.retain(|id| *id != auction_id);
+						});
+
+						let new_expiry = frame_system::Pallet::<T>::current_block_number()
+							.saturating_add(T::ExtendedLength::get());
+						auction_data.expiry_block = new_expiry;
+
+						// Add the new block number to the storage.
+						AuctionsExpiryBlock::<T>::append(new_expiry, auction_id);
+					}
+
+					Self::deposit_event(Event::<T>::Bid { who: id, auction_id, price });
+
+					Ok(())
+				} else {
+					Err(Error::<T>::InvalidAuctionId.into())
+				}
+			})
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::cancel_auction())]
+		pub fn cancel_auction(origin: OriginFor<T>, auction_id: AuctionId) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// todo read the storage auctions
+			let auction_data =
+				Auctions::<T>::get(auction_id).ok_or(Error::<T>::InvalidAuctionId)?;
+
+			// Ensure the nft is belong to the correct owner.
+			T::NftManager::ensure_nft_is_valid(&id, auction_data.nft_id)?;
```

Comment:
This check checks if the NFT exists and belong to a user. You should rename this